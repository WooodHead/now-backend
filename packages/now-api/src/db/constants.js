/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  DEVICE: 'now_device',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
