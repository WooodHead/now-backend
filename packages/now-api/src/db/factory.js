/* eslint-disable import/no-extraneous-dependencies  */
import { Factory } from 'rosie';
import uuid from 'uuid/v4';
import { address, name, date, internet, random, lorem, hacker } from 'faker';
import randomEmoji from 'random-emoji';
import { LocalDate, ZoneId, LocalDateTime } from 'js-joda';
import {
  EVENT_INVITE_TYPE,
  APP_INVITE_TYPE,
} from '../schema/resolvers/Invitation';

const md5 = () =>
  random
    .hexaDecimal(32)
    .toLowerCase()
    .substring(2);

Factory.define('user')
  .attr('id', uuid)
  .attr('auth0Id', () => random.alphaNumeric(24))
  .attr('email', internet.email)
  .attr('firstName', name.firstName)
  .attr('lastName', name.lastName)
  .attr('bio', lorem.paragraph)
  .attr('location', () => `${address.city()}, ${address.stateAbbr()}`)
  .attr('preferences', {})
  .attr('birthday', () =>
    LocalDate.now()
      .minusYears(20)
      .toString()
  )
  .attr('photoId', md5)
  .attr(
    'photoPreview',
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAAAXNSR0IArs4c6QAAAmRJREFUWAntmN1Kw0AQhU/8Fwv1F0FEKCgV9UbvvKkPIFJ8AB/C1/BZfAdF9EpBFEUEBQUp0l6opRVEi/Y4LIlx006aWHuRgTHpZHbmy9lssuh81g0dbF0dzPaNlgBGnaFEwUTBqAqY8e/vwM4OUCiYiO7Ytmfw+Bi4vRXXoUlW2wD396VhqRQGD2gL4N0dcH/fwYBGPSIWiwKq/atSsFbTlvudV60CJyduPOwU97hD7WcfH8D2NpBOA0tLwOIikM0CfX32fH/06AjgCjb29ga8vEg9E2t0bArYU8/gfod3vrcnztjcnMASeGoquMXBwe9rrMUb1lhTQBYZGwMeH91yVPXqSnx3FxgZEVgqPD8PDA5K7uWl3Jg7Us4IODvrj9p/twToL/X0BBweind3A5mMgJ6e+jPld5iFogIcH7c3skW5oG5uxG3XGQuzUFSrOAxgEJQ3Hjsgn8E4LcwU/4uCr68A348aUwHGrSDBtCqqAFMpYGBAc7/6HO1zqAJk27hVjB0waCVPTAC5HDA6qlePmdopVr0HWTBIwa0t+TYz5+EBODsTN9srxm2mVVANaFNwYcGFI8T0tPj6OvD8LKDn58D19c8NA3P/HNBxgM1NtrLb8DCwtibOHQy/y1T34gKoVIByWaB7e+3jTVStoH+KV1aAmRlTpvGxvx9gPp07I34KqWAzOFZVA3qnuKu+9vP5xlBBV6k8t2p0jalfM3wPDg1JydVVYHJSUz56jhqQragip2VjI3pjbQX1FLMgd8/Ly7JB1TaImuck/8CMKGGoZzBir5aGJ4AtyeYZlCjoEaOl0y+Le57bWRVobwAAAABJRU5ErkJggg=='
  );

Factory.define('device')
  .option('user', () => ({ id: uuid() }))
  .attr('token', uuid)
  .attr('type', () => random.arrayElement(['ios', 'android']))
  .attr('model', random.randomWord)
  .attr('userId', ['user'], ({ id }) => id);

Factory.define('rsvp')
  .option('user', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('action', 'add')
  .attr('userId', ['user'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('appInvite')
  .option('inviter', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', APP_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id);

Factory.define('eventInvite')
  .option('inviter', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', EVENT_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('activity')
  .attr('id', uuid)
  .attr('activityDate', () =>
    date
      .soon(1)
      .toISOString()
      .substring(0, 10)
  )
  .attr('title', () => `${hacker.ingverb()} for ${hacker.noun()}`)
  .attr('emoji', () => randomEmoji.random({ count: 1 })[0].character)
  .attr('description', hacker.phrase);

Factory.define('todayActivity')
  .extend('activity')
  .attr('activityDate', () => new Date().toISOString().substring(0, 10));

Factory.define('event')
  .option('activity', () => ({
    id: uuid(),
    activityDate: date
      .soon(1)
      .toISOString()
      .substring(0, 10),
  }))
  .option('location', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('activityId', ['activity'], ({ id }) => id)
  .attr('locationId', ['location'], ({ id }) => id)
  .attr('limit', () => random.number({ max: 20, min: 5 }))
  .attr('time', ['activity'], ({ activityDate: activityDateString }) => {
    const activityDate = LocalDate.parse(activityDateString);
    return date
      .between(
        activityDate.atStartOfDay(),
        activityDate.atTime4(23, 59, 59, 999)
      )
      .toISOString();
  })
  .attr('timezone', () => random.arrayElement(ZoneId.getAvailableZoneIds()));

Factory.define('location')
  .attr('id', uuid)
  .attr('name', hacker.ingverb)
  .attr('address', address.streetAddress)
  .attr('city', address.city)
  .attr('state', address.stateAbbr)
  .attr('postalCode', address.zipCode)
  .attr('country', address.countryCode)
  .attr('neighborhood', hacker.noun)
  .attr('foursquareVenueId', md5)
  .attr(
    'location',
    () => `SRID=4326;POINT(${address.longitude()} ${address.latitude()})`
  );

Factory.define('message')
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('ts', () => date.recent(1).getTime())
  .attr('text', hacker.phrase)
  .attr('eventId', ['event'], ({ id }) => id)
  .attr('userId', ['user'], ({ id }) => id);

export default Factory;
