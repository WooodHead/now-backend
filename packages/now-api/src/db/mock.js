/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';

import schema from '../schema';
import loaders from './loaders';

export const USER_ID = 'aea68a98-591c-11e8-9cdb-872a15ebfd30';
export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  loadMember: jest.fn(),
};

const apolloCache = new InMemoryCache();

let isAdmin = false;

export const setAdmin = b => {
  isAdmin = b;
};

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({
    schema,
    context: () => ({
      token: null,
      user: { id: USER_ID },
      loaders: loaders({ currentUserId: USER_ID }),
      scopes: isAdmin ? ['admin'] : [],
    }),
  }),
});

export const newUserClient = (currentUserAuth0Id, currentUserId) =>
  new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({
      schema,
      context: {
        token: null,
        currentUserAuth0Id,
        loaders: loaders({ currentUserId }),
      },
    }),
  });
