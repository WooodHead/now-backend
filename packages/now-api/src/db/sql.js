import knex from 'knex';
import { setPgTypes } from './types';
import knexConfig from '../../knexfile';

const config = knexConfig[process.env.NODE_ENV];
const sql = knex(config);

setPgTypes(sql);

export default sql;
export const connection = config.connection; // eslint-disable-line prefer-destructuring

export const genRandomUuid = () => sql.raw('gen_random_uuid()');
export const makepoint = (x, y) => sql.raw('st_makepoint(?, ?)', x, y);
export const now = () => sql.raw('now()');
