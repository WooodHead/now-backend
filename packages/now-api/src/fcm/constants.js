/* eslint-disable import/prefer-default-export */
export const NEW_EVENTS_TOPIC = 'new-events';
