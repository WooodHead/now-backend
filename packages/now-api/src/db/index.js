import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { rsvpId } from '../schema/util';
import { TABLES } from './constants';
import { isDev } from '../util';

const dynamoOpts = {};
if (isDev()) {
  dynamoOpts.logger = console;
}

const dynamoDb = new AWS.DynamoDB.DocumentClient(dynamoOpts);

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);

export const put = (table, item, condition = undefined) =>
  pPut({
    TableName: table,
    Item: item,
    ConditionExpression: condition,
  });

export const update = (
  table,
  key,
  expr,
  values,
  names = undefined,
  condition = undefined
) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ConditionExpression: condition,
    ReturnValues: 'ALL_NEW',
  });

export const updateDynamic = (table, key, values, condition = undefined) => {
  const expressions = [];
  const expressionNames = {};
  const expressionValues = {};
  Object.entries(values).forEach(([k, v], i) => {
    const attrName = `#f${i}`;
    const fieldName = `:f${i}`;
    expressions.push(`${attrName}=${fieldName}`);
    expressionNames[attrName] = k;
    expressionValues[fieldName] = v;
  });

  return update(
    table,
    key,
    `set ${expressions.join(', ')}`,
    expressionValues,
    expressionNames,
    condition
  );
};

export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

export const getUserRsvpByEvent = (userId, eventId) =>
  get(TABLES.RSVP, { id: rsvpId(eventId, userId) });

export const getActivity = id => get(TABLES.ACTIVITY, { id });

export const getEvent = id =>
  pQuery({
    TableName: TABLES.EVENT,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': id },
    IndexName: 'id-index',
  }).then(response => response.Items[0]);

export { TABLES };
