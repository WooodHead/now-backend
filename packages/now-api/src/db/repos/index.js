// @flow
import type knex from 'knex';

import defaultSql from '../sql';
import { SQL_TABLES } from '../constants';
import { putInOrder } from '../../util';

type flat = string | number;

class Repo {
  constructor(table: string, sql: knex = defaultSql) {
    this.table = table;
    this.sql = sql;
  }

  get = (query: mixed = {}) =>
    this.sql(this.table)
      .where(query)
      .first();

  all = (query: mixed = {}) => this.sql(this.table).where(query);

  insert = (obj: mixed) => this.sql(this.table).insert(obj);

  update = ({ id, ...otherFields }: { id: flat }) =>
    this.sql(this.table)
      .where({ id })
      .update(otherFields);

  batch = (ids: Array<flat>) =>
    this.sql(this.table) // TODO: log batch savings
      .whereIn('id', ids)
      .then(batch => putInOrder(batch, ids));

  delete = (query: mixed) =>
    this.sql(this.table)
      .where(query)
      .delete();

  byId = (id: flat) => this.get({ id });

  withTransaction = (trx: ?knex): Repo =>
    trx ? new Repo(this.table, trx) : this;

  sql: knex;
  table: string;
}

export const User = new Repo(SQL_TABLES.USERS);
export const Device = new Repo(SQL_TABLES.DEVICES);
export const Activity = new Repo(SQL_TABLES.ACTIVITIES);
export const Event = new Repo(SQL_TABLES.EVENTS);
export const Location = new Repo(SQL_TABLES.LOCATIONS);
export const Rsvp = new Repo(SQL_TABLES.RSVPS);
export const RsvpLog = new Repo(SQL_TABLES.RSVP_LOG);
export const Message = new Repo(SQL_TABLES.MESSAGES);
export const Invitation = new Repo(SQL_TABLES.INVITATIONS);
export const EventUserMetadata = new Repo(SQL_TABLES.EVENT_USER_METADATA);
export const ServerMessages = new Repo(SQL_TABLES.SERVER_MESSAGES);
