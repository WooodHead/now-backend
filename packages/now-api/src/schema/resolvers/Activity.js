import uuid from 'uuid';
import { LocalDateTime, LocalTime, ZoneId, LocalDate } from 'js-joda';

import { sqlPaginatify } from '../util';
import { Activity, Event } from '../../db/repos';
import sql from '../../db/sql';
import { notifyEventChange } from './Event';

export const NYC_TZ = ZoneId.of('America/New_York');
export const AVAILABILITY_HOUR = LocalTime.parse('21:00');
export const EARLY_AVAILABILITY_HOUR = LocalTime.parse('20:00');

export const getToday = () => {
  const now = LocalDateTime.now(NYC_TZ);
  return (now.toLocalTime().isBefore(AVAILABILITY_HOUR)
    ? now.toLocalDate()
    : now.toLocalDate().plusDays(1)
  ).toString();
};

const allActivities = (root, { input, orderBy = 'id' }) =>
  sqlPaginatify(orderBy, Activity.all({}), input);

const manyActivities = (root, { ids }, { loaders }) =>
  loaders.activities.loadMany(ids);

const activityQuery = (root, { id }, { loaders }) =>
  loaders.activities.load(id);

const todayActivity = () => Activity.get({ activityDate: getToday() });

export const queries = {
  todayActivity,
  activity: activityQuery,
  allActivities,
  manyActivities,
};

export const getEvents = ({ id }, { first, last, after, before }) =>
  sqlPaginatify('id', Event.all({ activityId: id }), {
    first,
    last,
    after,
    before,
  });

export const resolvers = {
  events: getEvents,
  generallyAvailableAt: () =>
    LocalDate.now()
      .atTime(AVAILABILITY_HOUR)
      .atZone(NYC_TZ),
};

const createActivity = (
  root,
  { input: { title, description, activityDate, emoji, pushNotification } },
  { loaders }
) => {
  const newId = uuid.v1();
  const newActivity = {
    id: newId,
    title,
    pushNotification,
    description,
    activityDate: activityDate.toString(),
    createdAt: sql.raw('now()'),
    updatedAt: sql.raw('now()'),
    emoji,
  };

  loaders.activities.clear(newId);

  return Activity.insert(newActivity).then(() => ({
    activity: loaders.activities.load(newId),
  }));
};

const updateActivity = async (
  root,
  { input: { id, title, description, activityDate, emoji, pushNotification } },
  { loaders }
) => {
  const updatedActivity = {
    id,
    title,
    pushNotification,
    description,
    activityDate: activityDate.toString(),
    updatedAt: sql.raw('now()'),
    emoji,
  };

  loaders.activities.clear(id);

  await Activity.update(updatedActivity);
  const activityEvents = await Event.all({ activityId: id });

  activityEvents.forEach(({ id: eventId }) => {
    notifyEventChange(eventId);
  });

  return {
    activity: loaders.activities.load(id),
  };
};

export const mutations = { createActivity, updateActivity };
