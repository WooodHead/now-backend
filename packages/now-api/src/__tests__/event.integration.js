import gql from 'graphql-tag';
import { omit, sortBy } from 'lodash';
import uuid from 'uuid/v4';
import { LocalDate } from 'js-joda';

import { client, setAdmin, USER_ID } from '../db/mock';
import { SQL_TABLES } from '../db/constants';
import sql from '../db/sql';
import factory from '../db/factory';
import { EventUserMetadata } from '../db/repos';
import { mockNow, restoreNow } from '../../testutils/date';
import * as Activity from '../schema/resolvers/Activity';
import { createRsvp } from '../schema/resolvers/Rsvp';

const activity = factory.build('activity');
const location = factory.build('location');
const events = factory.buildList('event', 5, {}, { activity, location });
const event = events[0];
event.time = '2018-05-30 20:00:00+00';
event.timezone = 'Asia/Calcutta';

const truncateTables = () =>
  Promise.all([
    sql(SQL_TABLES.ACTIVITIES).truncate(),
    sql(SQL_TABLES.EVENTS).truncate(),
    sql(SQL_TABLES.LOCATIONS).truncate(),
  ]);

beforeEach(() =>
  truncateTables().then(() =>
    Promise.all([
      sql(SQL_TABLES.ACTIVITIES).insert(activity),
      sql(SQL_TABLES.EVENTS).insert(events),
      sql(SQL_TABLES.LOCATIONS).insert(location),
    ])
  ));
afterEach(() => {
  setAdmin(false);
  return truncateTables();
});

describe('Event', () => {
  it('return allEvents', async () => {
    setAdmin(true);
    const results = client.query({
      query: gql`
        {
          allEvents {
            edges {
              node {
                id
                limit
                activity {
                  id
                  activityDate
                  title
                  description
                  createdAt
                  updatedAt
                }
                rsvps {
                  edges {
                    node {
                      event {
                        id
                      }
                      updatedAt
                    }
                  }
                }
                createdAt
                updatedAt
              }
            }
          }
        }
      `,
    });
    const { data } = await results;
    expect(data.allEvents).toMatchObject({
      edges: expect.arrayContaining(
        events.map(e =>
          expect.objectContaining({
            __typename: 'RootEventsEdge',
            [Symbol('id')]: expect.anything(),
            node: expect.objectContaining({
              __typename: 'Event',
              ...omit(e, [
                'locationId',
                'activityId',
                'timezone',
                'duration',
                'visibleAt',
                'time',
                'going',
              ]),
            }),
          })
        )
      ),
    });
  });

  describe('events visible to user', () => {
    let eventToday;
    let eventTomorrow;
    let eventLastWeek;
    beforeEach(async () => {
      await sql(SQL_TABLES.EVENTS).truncate();
      const today = LocalDate.now();
      eventToday = factory.build(
        'event',
        {
          time: today.atTime(14, 0).toString(),
          timezone: Activity.NYC_TZ.id(),
          visibleAt: today
            .minusDays(1)
            .atTime(14, 0)
            .toString(),
        },
        { activity, location }
      );
      eventTomorrow = factory.build(
        'event',
        {
          time: today
            .plusDays(1)
            .atTime(14, 0)
            .toString(),
          timezone: Activity.NYC_TZ.id(),
          visibleAt: today.atTime(14, 0).toString(),
        },
        { activity, location }
      );
      eventLastWeek = factory.build(
        'event',
        {
          time: today
            .minusDays(7)
            .atTime(14, 0)
            .toString(),
          timezone: Activity.NYC_TZ.id(),
          visibleAt: today
            .minusDays(8)
            .atTime(14, 0)
            .toString(),
        },
        { activity, location }
      );
      const eventNotPublished = factory.build(
        'event',
        {
          time: today.atTime(14, 0).toString(),
          timezone: Activity.NYC_TZ.id(),
          visibleAt: null,
        },
        { activity, location }
      );
      await sql(SQL_TABLES.EVENTS).insert([
        eventLastWeek,
        eventToday,
        eventTomorrow,
        eventNotPublished,
      ]);
    });
    afterEach(() => {
      restoreNow();
      return sql(SQL_TABLES.EVENTS).truncate();
    });
    it('before early availability', async () => {
      mockNow(
        LocalDate.now()
          .atTime(14, 0)
          .toString()
      );
      const results = client.query({
        fetchPolicy: 'network-only',
        query: gql`
          {
            events {
              edges {
                node {
                  id
                }
              }
            }
          }
        `,
      });

      const { data } = await results;

      expect(data.events.edges.map(({ node }) => node.id)).toEqual([
        eventToday.id,
        eventTomorrow.id,
      ]);
    });
    it('after early availability', async () => {
      mockNow(
        LocalDate.now()
          .atTime(20, 0)
          .atZone(Activity.NYC_TZ)
          .withFixedOffsetZone()
          .toString()
      );
      const results = client.query({
        fetchPolicy: 'network-only',
        query: gql`
          {
            events {
              edges {
                node {
                  id
                }
              }
            }
          }
        `,
      });
      const { data } = await results;

      expect(data.events.edges.map(({ node }) => node.id)).toEqual([
        eventToday.id,
        eventTomorrow.id,
      ]);
    });
    it('includePast', async () => {
      mockNow(
        LocalDate.now()
          .atTime(20, 0)
          .atZone(Activity.NYC_TZ)
          .withFixedOffsetZone()
          .toString()
      );
      const results = client.query({
        fetchPolicy: 'network-only',
        query: gql`
          query getEvents($includePast: Boolean) {
            events(includePast: $includePast) {
              edges {
                node {
                  id
                }
              }
            }
          }
        `,
        variables: {
          includePast: true,
        },
      });
      const { data } = await results;

      expect(data.events.edges.map(({ node }) => node.id)).toEqual([
        eventLastWeek.id,
        eventToday.id,
        eventTomorrow.id,
      ]);
    });
  });

  it('return event', async () => {
    const results = await client.query({
      query: gql`
        query getEvent($id: ID!) {
          event(id: $id) {
            id
            limit
            activity {
              id
              title
              emoji
              activityDate
              description
            }
            location {
              id
              foursquareVenueId
              lat
              lng
              address
              name
              crossStreet
              city
              state
              postalCode
              country
              neighborhood
            }
            rsvps {
              edges {
                node {
                  event {
                    id
                  }
                  updatedAt
                }
              }
            }
          }
        }
      `,
      variables: { id: event.id },
    });
    const { data } = results;
    expect(data).toMatchObject({
      event: {
        __typename: 'Event',
        ...omit(event, [
          'locationId',
          'activityId',
          'timezone',
          'duration',
          'visibleAt',
          'time',
          'going',
        ]),
        activity,
      },
    });
  });

  it('deals with time', async () => {
    const { id } = event;
    const result = await client.query({
      query: gql`
        query getEventTime($id: ID!) {
          event(id: $id) {
            id
            time
          }
        }
      `,
      variables: { id },
    });
    const foundTime = result.data.event.time;
    expect(foundTime).toEqual('2018-05-31T01:30+05:30');
  });

  describe('message counting', () => {
    afterEach(() =>
      Promise.all([
        sql(SQL_TABLES.EVENT_USER_METADATA).truncate(),
        sql(SQL_TABLES.MESSAGES).truncate(),
      ]));
    it('defaults to 0', async () => {
      const results = await client.query({
        query: gql`
          query getEvent($id: ID!) {
            event(id: $id) {
              id
              messages {
                count
                unreadCount
              }
            }
            unreadMessagesCount
          }
        `,
        variables: { id: event.id },
        fetchPolicy: 'network-only',
      });
      const { data } = results;
      expect(data.event.messages).toMatchObject({
        unreadCount: 0,
        count: 0,
      });
      expect(data.unreadMessagesCount).toEqual(0);
    });
    it('unread matches count when no row created', async () => {
      const messages = factory.buildList(
        'message',
        10,
        {},
        { event, user: { id: USER_ID } }
      );
      await sql(SQL_TABLES.MESSAGES).insert(messages);
      const results = await client.query({
        query: gql`
          query getEvent($id: ID!) {
            event(id: $id) {
              id
              messages {
                count
                unreadCount
              }
            }
            unreadMessagesCount
          }
        `,
        variables: {
          id: event.id,
        },
        fetchPolicy: 'network-only',
      });
      const { data } = results;
      expect(data.event.messages).toMatchObject({
        unreadCount: 10,
        count: 10,
      });
      expect(data.unreadMessagesCount).toEqual(0);
    });
    it('unread matches count when RSVP', async () => {
      await sql.transaction(trx =>
        createRsvp(
          trx,
          { userId: USER_ID, eventId: event.id, ignoreVisible: true },
          'add'
        )
      );
      const messages = factory.buildList(
        'message',
        10,
        {},
        { event, user: { id: USER_ID } }
      );
      await sql(SQL_TABLES.MESSAGES).insert(messages);
      const { data } = await client.query({
        query: gql`
          query unreadMessagesCount {
            unreadMessagesCount
          }
        `,
        fetchPolicy: 'network-only',
      });
      expect(data.unreadMessagesCount).toEqual(10);
    });
    it('unread correct when marked read', async () => {
      const messages = factory.buildList(
        'message',
        10,
        {},
        { event, user: { id: USER_ID } }
      );
      await sql(SQL_TABLES.MESSAGES).insert(messages);

      const sortedMessages = sortBy(messages, 'ts');
      const unreadTs = sortedMessages[3].ts;

      await sql(SQL_TABLES.EVENT_USER_METADATA).insert({
        id: uuid(),
        lastReadTs: unreadTs,
        eventId: event.id,
        userId: USER_ID,
      });

      const results = await client.query({
        query: gql`
          query getEvent($id: ID!) {
            event(id: $id) {
              id
              messages {
                count
                unreadCount
              }
            }
          }
        `,
        variables: {
          id: event.id,
        },
        fetchPolicy: 'network-only',
      });
      const { data } = results;
      expect(data.event.messages).toMatchObject({
        unreadCount: 6,
        count: 10,
      });
    });
  });

  describe('markEventChatRead', () => {
    it('inserts new read time', async () => {
      const results = await client.mutate({
        mutation: gql`
          mutation mark($input: MarkEventChatReadInput!) {
            markEventChatRead(input: $input) {
              event {
                id
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, ts: '123456' } },
      });
      const { data } = results;
      expect(data).toMatchObject({
        markEventChatRead: {
          __typename: 'MarkEventChatReadPayload',
          event: {
            __typename: 'Event',
            id: event.id,
          },
        },
      });

      const dbMetadata = await EventUserMetadata.get({
        eventId: event.id,
        userId: USER_ID,
      });

      expect(dbMetadata).toMatchObject({
        userId: USER_ID,
        eventId: event.id,
        lastReadTs: '123456',
      });
    });

    it('requires ts as number', async () => {
      const results = client.mutate({
        mutation: gql`
          mutation mark($input: MarkEventChatReadInput!) {
            markEventChatRead(input: $input) {
              event {
                id
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, ts: 'not a number' } },
        errorPolicy: 'none',
      });

      expect.assertions(1);
      return expect(results).rejects.toEqual(
        new Error('GraphQL error: ts must be an integer as a string')
      );
    });

    it('updates existing read time', async () => {
      await client.mutate({
        mutation: gql`
          mutation mark($input: MarkEventChatReadInput!) {
            markEventChatRead(input: $input) {
              event {
                id
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, ts: '123456' } },
      });
      const results = await client.mutate({
        mutation: gql`
          mutation mark($input: MarkEventChatReadInput!) {
            markEventChatRead(input: $input) {
              event {
                id
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, ts: '654321' } },
      });
      const { data } = results;
      expect(data).toMatchObject({
        markEventChatRead: {
          __typename: 'MarkEventChatReadPayload',
          event: {
            __typename: 'Event',
            id: event.id,
          },
        },
      });

      const dbMetadata = await EventUserMetadata.get({
        eventId: event.id,
        userId: USER_ID,
      });

      expect(dbMetadata).toMatchObject({
        userId: USER_ID,
        eventId: event.id,
        lastReadTs: '654321',
      });
    });
  });
});
