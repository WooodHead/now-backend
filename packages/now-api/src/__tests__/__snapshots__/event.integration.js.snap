// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event createEvent rejects an invalid community id 1`] = `[Error: GraphQL error: No Such Community]`;
